segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/rolling/RollingStorage.java

```diff
@@ -472,13 +476,20 @@ public void truncate(SegmentHandle handle, long truncationOffset) throws StreamS
         Preconditions.checkArgument(truncationOffset >= 0 && truncationOffset <= h.length(),
                 "truncationOffset must be non-negative and at most the length of the Segment.");
         val last = h.lastChunk();
-        if (last != null && canTruncate(last, truncationOffset)) {
-            // If we were asked to truncate the entire Segment, then rollover at this point so we can delete all existing
-            // data.
+        if (last != null && canTruncate(last, truncationOffset) && !h.isSealed()) {
+            // If we were asked to truncate the entire (non-sealed) Segment, then rollover at this point so we can delete
+            // all existing data.
             rollover(h);
+
+            // We are free to delete all chunks.
+            deleteChunks(h, s -> canTruncate(s, truncationOffset));
+        } else {
+            // Either we were asked not to truncate the whole segment, or we were, and the Segment is sealed. If the latter,
+            // then the Header is also sealed, we could not have done a quick rollover; as such we have no option but to
+            // preserve the last chunk so that we can recalculate the length of the Segment if we need it again.
+            deleteChunks(h, s -> canTruncate(s, truncationOffset) && s.getLastOffset() < h.length());
```

Comment:
I'm not sure I understand the predicate here because of the following.

`canTruncate` only evaluates to `true` in the case that the segment is entirely under the truncation offset:

```
return segmentChunk.getStartOffset() < truncationOffset
                && segmentChunk.getLastOffset() <= truncationOffset;
```

We have already verified above that: `truncationOffset <= h.length()` (Line 476).  Consequently, the only case in which `canTruncate` evaluates to `true` and `s.getLastOffset() < h.length()` to `false`, which prevents the chunk from being deleted, is the one in which ` s.getLastOffset() = truncationOffset = h.length()`. But, if these three values are all equal, then I think we are saying that we want to delete the whole segment, no? If this analysis is right, then using `canTruncate` alone should be sufficient.

Could you help me sort out my confusion? 

Another related but separate question is where we update `h.length()` upon truncation. I understand that is not related to this bug, it is for my own understanding.

Reply:
You are correct, but only in the case when `h.isSealed() == false` (segment is not sealed). As you can see from this `if-else` block, you are commenting on the `else` part, which means this applies only for a `sealed` segment. 

What I am doing with this condition is, if we were asked to truncate on the last offset (otherwise this would not be a problem), then delete all chunks but the last. So, if `truncationOffset == h.length()`,  the condition `canTruncate(s, truncationOffset) && s.getLastOffset() < h.length()` is:
- `true` for all chunks but the last chunk
- `false` for the last chunk (because `canTruncate == true` but the second part evaluates to `false`.

This is the intended behavior, which is fixed with this PR.

`h.length()` is never updated during truncation. As per our design, truncating a segment does not change its length, otherwise we'd have a mess with offsets remapping to smaller offsets, thus leading to inconsistent writes. The length is only modified upon a successful write or concat operation. 

There is no `StartOffset` concept in Tier2 (that is stored in the Segment's Metadata in the Store, in an upstream layer). If you issue a truncate at offset N in Tier2, then the `RollingStorage` will delete everyting before offset M<=N in the segment, where M is the last rolling point for that segment. The range [M,N) will still be accessible in Tier2, but it will not via the Store's API (due to the latter enforcing the StartOffset, which it can stash away into the metadata).

Reply:
Thanks for the clarification @andreipaduroiu . The thing I'm missing now is that I don't understand why we want to keep the last chunk and header around in the case that `truncationOffset == h.length()`. Aren't we saying in this case that we want to delete any remaining data in the segment? 

There is probably something really basic that is escaping my attention.

Reply:
Hmm, I hoped the comment right above this line would be doing a better job at explaining the problem (also explained in the last bullet in the PR description). 

At this point the Header file is sealed, wo we cannot make any more modifications to it. The header file only has pointers to the individual chunks. In order to calculate the length of the segment, it needs information from Tier2 for the last chunk (starting offset + length). If we delete the last chunk, then we will have no way of re-acquiring this information. For non-sealed segments, it's simpler, since we just force a rollover at that point and create an empty chunk. but for sealed ones, we cannot do that since we cannot modify the header file.

Reply:
> the Header file is sealed, wo we cannot make any more modifications to it

Ok.

>  In order to calculate the length of the segment, it needs information from Tier2 for the last chunk (starting offset + length).

Ok.

> If we delete the last chunk, then we will have no way of re-acquiring this information.

Ok.

I don't think any of that answers my question, though:

In the case that `truncationOffset == h.length()`,  aren't we saying that we want to delete any remaining data in the segment? My interpretation is that we are essentially getting rid of the segment in this case, and consequently, there is no reason to keep anything around. Again, I'm focusing on the `truncationOffset == h.length()` scenario.

Reply:
No. Just because we truncate a sealed segment at the end doesn't mean we want to delete it. If we want to delete it then the controller or client needs to request that explicitly.